src/hotspot/share/opto/reachability.cpp

```diff
@@ -323,90 +211,85 @@ bool PhaseIdealLoop::optimize_reachability_fences() {
 
   assert(OptimizeReachabilityFences, "required");
 
+  // ResourceMark rm; // NB! not safe because insert_rf may trigger _idom reallocation
   Unique_Node_List redundant_rfs;
-  find_redundant_rfs(redundant_rfs);
-
-  Node_List worklist;
+  GrowableArray<Pair<Node*,Node*>> worklist;
   for (int i = 0; i < C->reachability_fences_count(); i++) {
     ReachabilityFenceNode* rf = C->reachability_fence(i);
-    if (!redundant_rfs.member(rf)) {
-      // Move RFs out of counted loops when possible.
-      IdealLoopTree* lpt = get_loop(rf);
-      Node* referent = rf->referent();
-      Node* loop_exit = lpt->unique_loop_exit_or_null();
-      if (lpt->is_invariant(referent) && loop_exit != nullptr) {
-        // Switch to the outermost loop.
-        for (IdealLoopTree* outer_loop = lpt->_parent;
-             outer_loop->is_invariant(referent) && outer_loop->unique_loop_exit_or_null() != nullptr;
-             outer_loop = outer_loop->_parent) {
-          assert(is_member(outer_loop, rf), "");
-          loop_exit = outer_loop->unique_loop_exit_or_null();
-        }
-        assert(loop_exit != nullptr, "");
-        worklist.push(referent);
-        worklist.push(loop_exit);
-        redundant_rfs.push(rf);
+    assert(!rf->is_redundant(igvn()), "required");
```

Comment:
Why can we assume this? Is this guaranteed by IGVN?

Reply:
Yes, the assumption is that last IGVN pass removed all redundant RFs and the graph hasn't changed for new cases to occur.